File: Leetcode/dfs_recursion_backtracking/695_MaxAreaOfIsland/Solution.py
class Solution(object):
    def maxAreaOfIsland(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        def dfs(sr, sc):
            if (not (0 <= sr <len(grid) and 0 <= sc <len(grid[sr]))):
                return 0
            if grid[sr][sc]==0:
                return 0
            grid[sr][sc] = 0
            return 1+sum([dfs(sr + x, sc + y) for (x, y) in ((0, 1), (1, 0), (0, -1), (-1, 0))])
        res = 0
        for i in xrange(0, len(grid)):
            for j in xrange(0, len(grid[i])):
                res = max(dfs(i, j), res)
        return res

File: Leetcode/dfs_recursion_backtracking/695_MaxAreaOfIsland/Solution.cpp
class Solution {
private:
    int dfs(vector<vector<int>>& grid, int i, int j){
        if(i<0||i>=grid.size()||j<0||j>=grid[i].size() )
            return 0;
        if(grid[i][j]==0)
            return 0;
        grid[i][j]=0;
        return 1 + dfs(grid, i+1, j) + dfs(grid, i-1, j)
                + dfs(grid, i, j+1) + dfs(grid, i, j-1);
    }
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int res = 0;
        for(int i = 0; i<grid.size(); ++i){
            for(int j = 0; j<grid[i].size(); ++j)
                res = max(dfs(grid, i, j), res);
        }
        return res;
    }
};
